### PR TITLE
run CI tests also in osx environment

### PR DESCRIPTION
**Pull request by i110**
Run tests in osx too. In addition, introduce TravisCI cache feature to reduce buildtime.
One issue: `t/50access-log.t` randomly fails presumably due to VM's filesystem issue (specific in OSX). I tried to find the reasons and reasonable fixes (e.g. change TMPDIR to `/home/travis/tmp`), but I couldn't. So let it skip in TravisCI OSX environment.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,28 +1,68 @@
...
+        # install openssl-1.0.2
+        if [[ ! -e "${INSTALL_DIR}/lib/libssl.a" ]]; then
+          curl -L https://www.openssl.org/source/openssl-1.0.2h.tar.gz | tar xzf -
+          (cd openssl-1.0.2h && ./Configure darwin64-x86_64-cc --prefix=${INSTALL_DIR} no-shared && make && make install)
+        fi
+        # install zlib-1.2.8
+        if [[ ! -e "${INSTALL_DIR}/lib/libz.a" ]]; then
+          curl -L http://zlib.net/zlib-1.2.8.tar.gz | tar xzf -
+          (cd zlib-1.2.8 && ./configure --prefix=${INSTALL_DIR} --static && make && make install)
+        fi
```
Comment (kazuho) [inline .travis.yml]: Is there any reason for _not_ using openssl and zlib using homebrew, considering the fact that we install other packages provided by homebrew?

File: .travis.yml [hunk 2]
```diff
@@ -1,28 +1,68 @@
...
-  - $CXX --version
-  # for speed, pre-install deps installed in `before_script` section as ubuntu packages
-  - sudo apt-get install -qq cpanminus libipc-signal-perl liblist-moreutils-perl libwww-perl libio-socket-ssl-perl zlib1g-dev
+  - source ./misc/travis/export_env
+
+  # get latest version of cpanminus
+  - |
+    if [[ ! -e "${INSTALL_DIR}/bin/cpanm" ]]; then
+      curl -L https://cpanmin.us/ --create-dirs -o ${INSTALL_DIR}/bin/cpanm
+      chmod +x ${INSTALL_DIR}/bin/cpanm
```
Comment (kazuho) [inline .travis.yml]: Is there any reason to use a non-standard way of installing instead of running `curl -L https://cpanmin.us | perl - App::cpanminus`?

**Comment by kazuho** [comment]
Thank you very much for spending your time on this PR. I understand it's hard to tweak Travis, and my respect goes to what you have accomplished so far.

That said, in addition to the my two comments in-line let me discuss how we cache the binaries.

In the PR, whether if a binary should be built is determined by checking the existence of a library (i.e. `.a` file) or an executable. I would like to request a modification and a clarification to the method.

1) please distinguish the files between versions

As time evolves, it is likely that we would upgrade the dependencies (e.g. switch from openssl 1.0.2 to openssl 1.1.0). Such change happens per-branch.

In other words, in the future we would see our project use different versions of the dependencies, depending on the branch. For example, master branch could switch to using openssl 1.1.0 while 2.0 branch stays using openssl 1.0.2.

To me it seems like to me that the current approach does not take such possibility into consideration. Is my observation correct? If yes, please consider how to fix the issue.

2) please clarify how to rebuild a library

Sometimes we would need to rebuild the library. I would like to at least see a comment on how to trigger a rebuild.

File: .travis.yml (hunk 1 above)
Comment (i110) [inline .travis.yml]: I think it's better not to use homebrew where we can build from source, because
* as a premise, caching homebrew libraries seems to be heavy job (re-installing homebrew is required I believe)
* so, building from source once and reusing it is faster than installing via homebrew every time
* as additional bonus, by doing this, we can use the same versions in all environments, and don't have to care anymore how to install specific version which doesn't exist in homebrew repository

File: .travis.yml (hunk 2 above)
Comment (i110) [inline .travis.yml]: The reasons are:
* in my trial, the shell couldn't find cpanm executable
* I saw downloading occasionally failed due to cpanminus server issues, so I thought it's better to cache this too

**Comment by i110** [comment]
@kazuho Thank you the review.

> To me it seems like to me that the current approach does not take such possibility into consideration. Is my observation correct? If yes, please consider how to fix the issue.
We can clear the cache via TravisCI UI (`Caches` in the right-top `More options` menu). TravisCI have the caches for each PR, and each PR target branch, so if we want to upgrade libraries in other branches, simply clearing the cache first will work well, I believe,
But basically, I agree with you that distinguishing the versions is good idea, so I'll find the way. How about to rename the lib file like `~/local/lib/libuv.a.1.0.0` and create symlink as `~/local/lib/libuv.a` which points to the versioned one? Then we can check the cache by seeing the existence of the specific version lib file.

> Sometimes we would need to rebuild the library. I would like to at least see a comment on how to trigger a rebuild.
As mentioned above, we can clear the cache via TravisCI UI.

File: .travis.yml (hunk 1 above)
Comment (kazuho) [inline .travis.yml]: Thank you for the answer. Sounds convincing.

Then, could you please clarify why homebrew is used to install _some_ dependencies?

I am not suggesting that we should entirely abandon, I am just asking how we should decide between building from source and using homebrew, when adding a new library.

**Comment by i110** [comment]
@kazuho worked well!

File: .travis.yml [hunk 3]
```diff
@@ -1,28 +1,54 @@
...
-  # upgrade g++ and libstdc++ to build nghttp2
-  - sudo add-apt-repository --yes ppa:ubuntu-toolchain-r/test
-  - sudo apt-add-repository --yes ppa:smspillaz/cmake-2.8.12
-  - sudo apt-get --yes update
-  - sudo apt-get install --yes cmake cmake-data g++-4.8 libstdc++-4.8-dev php5-cgi wget
-  - if [ "$CXX" = "g++" ]; then export CXX="g++-4.8"; fi
-  - $CXX --version
-  # for speed, pre-install deps installed in `before_script` section as ubuntu packages
-  - sudo apt-get install -qq cpanminus libipc-signal-perl liblist-moreutils-perl libwww-perl libio-socket-ssl-perl zlib1g-dev
+  - source ./misc/travis/export_env
```
Comment (kazuho) [inline .travis.yml]: Should we better use `.` instead of `source` considering that the latter is a bash extension (is it?)?

Comment (kazuho) [inline .travis.yml]: Or, is there any reason to have the environment variables listed in a dedicated file considering the fact that `export_env` is only used once?

File: misc/travis/export_env [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#! /bin/sh
+export INSTALL_DIR=${HOME}/local
+export CPANM_DIR=${HOME}/cpanm
+export PATH=$(misc/travis/insert_path.pl ${INSTALL_DIR}/bin ${CPANM_DIR}/bin)
```
Comment (kazuho) [inline misc/travis/export_env]: Maybe use something like below would do without using a dedicated perl script?
```
export PATH=`echo -n "$PATH" | sed -E 's!(:/usr(/local/bin|/bin):)!:${INSTALL_DIR}/bin:${CPANM_DIR}/bin:$1!`
```

File: t/Util.pm [hunk 5]
```diff
@@ -239,6 +239,10 @@ sub prog_exists {
     system("which $prog > /dev/null 2>&1") == 0;
 }
 
+sub travis_os_name {
```
Comment (kazuho) [inline t/Util.pm]: A test should never depend on travis.

**Comment by kazuho** [comment]
Thank you for the tweaks.

I have left some comments in hope that we can make the CI simpler.

I also think that instead of JSON we should use separate file for storing the versions of each dependency. That way, you can write version check as simply as like follows instead of writing down a dedicated perl script.

```sh
if [ "`cat $VERSION_DIR/libuv 2> /dev/null`" = "1.0.0" ] ; then
    ...
```

**Comment by i110** [comment]
@kazuho I addressed the points you said.

File: .travis.yml [hunk 6]
```diff
@@ -33,17 +69,68 @@ before_script:
...
+          curl -L https://github.com/libuv/libuv/archive/v${LIBUV_VERSION}.tar.gz | tar xzf -
+          (cd libuv-${LIBUV_VERSION} && ./autogen.sh && ./configure --prefix=${INSTALL_DIR} && make && make install)
+          echo "${LIBUV_VERSION}" > ${CACHE_VERSION_DIR}/libuv
+        fi
+        ;;
+      "osx"   )
+        brew install --quiet libuv
+        brew link libuv
+        ;;
+    esac
```
Comment (kazuho) [inline .travis.yml]: Should we merge this to the OS-dependent installation of other modules done earlier?

File: .travis.yml [hunk 7]
```diff
@@ -33,17 +69,68 @@ before_script:
...
   - misc/install-perl-module.pl Test::Exception
+  # install libuv >= 1.0.0 (optionally required for building / testing libh2o)
+  - |
+    case "${TRAVIS_OS_NAME}" in
+      "linux" )
+        LIBUV_VERSION=1.0.0
+        if [ "`cat ${CACHE_VERSION_DIR}/libuv 2> /dev/null`" != "${LIBUV_VERSION}" ] ; then
+          curl -L https://github.com/libuv/libuv/archive/v${LIBUV_VERSION}.tar.gz | tar xzf -
+          (cd libuv-${LIBUV_VERSION} && ./autogen.sh && ./configure --prefix=${INSTALL_DIR} && make && make install)
+          echo "${LIBUV_VERSION}" > ${CACHE_VERSION_DIR}/libuv
```
Comment (kazuho) [inline .travis.yml]: You should not create a version file, unless the installation steps succeed (and this might also apply for the case blocks in other parts of the PR).

File: .travis.yml [hunk 8]
```diff
@@ -1,28 +1,64 @@
...
+        if [ "$CXX" = "g++" ]; then export CXX="g++-4.8"; fi
+        $CXX --version
+        # for speed, pre-install deps installed in `before_script` section as ubuntu packages
+        sudo apt-get install -qq libipc-signal-perl liblist-moreutils-perl libwww-perl libio-socket-ssl-perl zlib1g-dev
+        ;;
+      "osx"   )
+        brew update
+        brew tap homebrew/dupes
+        brew install --quiet openssl homebrew/dupes/zlib
+        brew link --force openssl
```
Comment (DomT4) [inline .travis.yml]: This won't work, FYI. It can cause security issues so it's blocked off.

You should either pass flags or prepend your `PATH` with the directory - Potentially `CMAKE_PREFIX_PATH` rather than `PATH` given `h2o` uses CMake.

File: .travis.yml [hunk 9]
```diff
@@ -1,28 +1,64 @@
...
+        sudo apt-get install -qq cmake cmake-data g++-4.8 libstdc++-4.8-dev php5-cgi wget
+        if [ "$CXX" = "g++" ]; then export CXX="g++-4.8"; fi
+        $CXX --version
+        # for speed, pre-install deps installed in `before_script` section as ubuntu packages
+        sudo apt-get install -qq libipc-signal-perl liblist-moreutils-perl libwww-perl libio-socket-ssl-perl zlib1g-dev
+        ;;
+      "osx"   )
+        brew update
+        brew tap homebrew/dupes
+        brew install --quiet openssl homebrew/dupes/zlib
```
Comment (DomT4) [inline .travis.yml]: Would be surprised if you need to lean on our `zlib`. Apple ships `zlib` themselves, and it is kept reasonably up-to-date as far as OS X's system utilities go.

File: .travis.yml (hunk 6 above)
Comment (i110) [inline .travis.yml]: Sure, I'll merge.
